### DISCUSSION TITLE
How to read data from int16_t queue

### DISCUSSION THREAD

**Discussion by hiteshzamad** [google/oboe#1235] [Q&A] [answered]
Problem :: Music not audible 
Android :: 16 
Device :: Samsung galaxy star pro

I am reading a wav file using fread in c++ as int16_t buffer array.
simultaneously i want to play the int16_t array using c++.
Is it a proper way of doing it ??

void Player::init_builder(uint16_t size_channel_t, uint32_t sample_rate) {
builder.setChannelCount(size_channel_t)
->setSampleRate(sample_rate)
->setDirection(oboe::Direction::Output)
->setContentType(oboe::ContentType::Music)
->setSharingMode(oboe::SharingMode::Exclusive)
->setSampleRateConversionQuality(oboe::SampleRateConversionQuality::Medium)
->setPerformanceMode(oboe::PerformanceMode::LowLatency)
->setFormat(oboe::AudioFormat::I16)
->setCallback(this);
}

int32_t Player::init_stream() {
return (int32_t) builder.openStream(audio_stream);
}

oboe::DataCallbackResult Player::onAudioReady(oboe::AudioStream *oboeStream, void *audioData, int32_t numFrames) {
if (resource->buffer_int_16 == nullptr){
resource->data_required = false;
return oboe::DataCallbackResult::Stop;
}
int r = (((int)resource->buffer_int_16->size() / 2) - (int)numFrames);
if (r == 0)
return oboe::DataCallbackResult::Continue;

if (r < 512)
    resource->data_required = true;
else
    resource->data_required = false;
auto *int_16_data = (int16_t*) (audioData);
int i;
for ( i = 0; i < numFrames && resource->buffer_int_16->size()/2 > 0; ++i) {
    for (int j = 0; j < 2; ++j) {
        int_16_data[i * 2 + j] = resource->buffer_int_16->front();
        resource->buffer_int_16->pop();
    }
}
return oboe::DataCallbackResult::Continue;
}

**Comment by dturner** [comment, marked as answer]
Code looks like it should work. Are you calling `AudioStream::requestStart()` anywhere to start the callbacks?

**Reply by hiteshzamad** [reply]
I got the error it is due to not filling all numframes . I forgot to put 0 where buffer size is less than numframes.

**Reply by hiteshzamad** [reply]
I have to stream music from internet and edit the buffer then i have to play the buffer using oboe. Do you know any library which can be use to load online audio buffer to android device?
